### PR TITLE
Multi root (vscode next)

### LINKED ISSUE #243: Multi-root Workspaces and Prettier
Hi,

I'm from the VS Code team and I'm reaching out with some additional resources for adopting multi-root workspaces.

But first let me say thank you for working on and maintaining [Prettier This for VS Code](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) which is one of the most popular extensions, we really appreciate the work you have done to make VS Code better.

The purpose of this issue is that we are very close to releasing a significant new feature for VS Code - Multi-root Workspaces - which is the [most requested](https://github.com/Microsoft/vscode/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request+sort%3Areactions-%2B1-desc) missing feature. If you have not heard of multi-root workspaces we have [described this in several of our recent updates](https://code.visualstudio.com/updates/v1_17#_preview-multi-root-workspaces).  Today it's only enabled in our insiders release but **our intent is to enable it for all users** in our [next update](https://github.com/Microsoft/vscode/issues/35875), targeted for release in the first week of November.

This is a pretty significant change and over the last four months we have updated pretty much all of VS Code to be Multi-root aware as well as updated all of our own extensions.  In the process of doing this we have created a set of [guidelines for Extension Authors](https://github.com/Microsoft/vscode/wiki/Extension-Authoring:-Adopting-Multi-Root-Workspace-APIs) to follow to ensure their extension can fully leverage the awesomeness that is Multi-root workspaces.

**Our Request:**

Having looked at your extension we think changes are required in how you:

* Make use of the (now deprecated) `workspace.rootPath` property, there is a proposed migration path and sample for this case.
* Contribute or change settings, specifically you need to review whether some of the settings can be applied on a resource (= file location) level instead of being global (= workspace level).

**How Can We Help/Resources:**

We realize we are asking for your help here - our goal is to make sure that your extensions works well with multi-root and continues to be loved by the community.  [The guide above outlines the changes that are required](https://github.com/Microsoft/vscode/wiki/Extension-Authoring:-Adopting-Multi-Root-Workspace-APIs) and we have a [samples](https://github.com/Microsoft/vscode/wiki/Extension-Authoring:-Adopting-Multi-Root-Workspace-APIs#samples) as well.

But we are sure you will have questions and we thought one way we could look to help is by getting a number of the team members (myself included) to be ready to respond to questions on a Slack Channel that we will dedicate to this issue.  To join that channel simply [follow this link](https://vscode-dev-community.slack.com/join/shared_invite/enQtMjIxOTgxNDE3NzM0LWU5M2ZiZDU1YjBlMzdlZjA2YjBjYzRhYTM5NTgzMTAxMjdiNWU0ZmQzYWI3MWU5N2Q1YjBiYmQ4MzY0NDE1MzY).

Again thank you for being part of the VS Code community - we appreciate everything you have done.  With the release of Multi-root Workspaces we will be shipping our most requested feature, together we can really help this feature light up for users.

Happy Coding!
Sean McBreen

// cc @bpasero @chrmarti

### PR DESCRIPTION
**Pull request by CiGit**
Adapt settings for resource usage. ie settings which are relevant to a file.

remove rootpath usage.

Will need to make test on an insider version. (Welcome testers !)

fix #243 

Edit: Made some tests on insider with 2 folders. Was working 😄

### CODE REVIEW

**Comment by felixfbecker** [comment]
Nice! Is there a way to enable prettier only per folder? I.e. a folder setting `"prettier.enable": true/false`. That would be especially important for multi-folder workspaces to not cause unwanted formatting changes

File: package.json [hunk 1]
```diff
@@ -19,7 +19,7 @@
     "url": "https://github.com/prettier/prettier-vscode/issues"
   },
   "engines": {
-    "vscode": "^1.15.0"
+    "vscode": "^1.17.0"
   },
   "categories": [
     "Formatters"
```
Comment (azz) [inline package.json]: Going to add `multi root ready`?

https://github.com/prettier/prettier-vscode/issues/243#issuecomment-340616573

Comment (CiGit) [reply package.json]: eventually yes.

Just wanted to really test it first.

**Comment by CiGit** [comment]
@felixfbecker This is not possible. Either you register a formatter for the entire instance or you don't.
However `.prettierignore` may do what you want.

**Comment by felixfbecker** [comment]
@CiGit 

> Either you register a formatter for the entire instance or you don't

Sorry I don't understand. Why is this not possible (but .prettierignore is)?

**Comment by CiGit** [comment]
I may have misunderstood you question. To me you wanted to be able to use any other formatter (at least the default one) on *other* folder. ie enable the extension per folder. This may be a feature from vscode.

If it's just a toggle, on/off, this may be possible, but I would be against it. #199 would already do it. Once implemented.

**Comment by azz** [comment]
If you're looking to turn off formatOnSave for a given workspace folder, I believe you can just put `"formatOnSave": false` in `.vscode/settings.json` in that folder and it will work.

**Comment by felixfbecker** [comment]
I think #199 would solve it. "Format document" needs to use a different formatter depending on the workspace folder.

**Comment by rothfels** [comment]
@felixfbecker afaict a formatter extension must register itself (and therefore override default formatters shipped w/ vscode) before we have specific information about the the workspaces in a multi-root setting. The `.prettierignore` option works to disable prettier formatting on a workspace root, but the prettier extension will have still been registered and overridden the default formatter (even though it'sa noop). I don't think there's currently a way to dynamically register a formatter extension based on properties/config of the workspace root.  This is unfortunate because (especially in a multi-root context) I imagine many people will run into an issue lacking the ability to run multiple formatters for the same language id simultaneously.

https://github.com/Microsoft/vscode/issues/15611 might be the appropriate solution

**Comment by CiGit** [comment]
Ready to review! I've nothing more to test or add

**Pull request merged**